File: src/main/java/org/luxons/sevenwonders/game/data/definitions/CardDefinition.java
package org.luxons.sevenwonders.game.data.definitions;

import java.util.List;
import java.util.Map;

import org.luxons.sevenwonders.game.Settings;
import org.luxons.sevenwonders.game.cards.Card;
import org.luxons.sevenwonders.game.cards.Color;
import org.luxons.sevenwonders.game.cards.Requirements;

public class CardDefinition implements Definition<Card> {

    private String name;

    private Color color;

    private Requirements requirements;

    private EffectsDefinition effect;

    private String chainParent;

    private List<String> chainChildren;

    private Map<Integer, Integer> countPerNbPlayer;

    private String image;

    @Override
    public Card create(Settings settings) {
        return new Card(name, color, requirements, effect.create(settings), chainParent, chainChildren, image);
    }

    Map<Integer, Integer> getCountPerNbPlayer() {
        return countPerNbPlayer;
    }
}

File: game-engine/src/main/kotlin/org/luxons/sevenwonders/game/data/definitions/CardDefinition.kt
package org.luxons.sevenwonders.game.data.definitions

import org.luxons.sevenwonders.game.cards.Card
import org.luxons.sevenwonders.game.cards.CardBack
import org.luxons.sevenwonders.game.cards.Color
import org.luxons.sevenwonders.game.cards.Requirements

internal class CardDefinition(
    private val name: String,
    private val color: Color,
    private val requirements: Requirements = Requirements(),
    private val effect: EffectsDefinition,
    private val chainParent: String? = null,
    private val chainChildren: List<String> = emptyList(),
    private val image: String,
    private val countPerNbPlayer: Map<Int, Int>
) {
    fun create(back: CardBack, nbPlayers: Int): List<Card> = List(countPerNbPlayer[nbPlayers] ?: 0) { create(back) }

    fun create(back: CardBack): Card =
        Card(name, color, requirements, effect.create(), chainParent, chainChildren, image, back)
}
